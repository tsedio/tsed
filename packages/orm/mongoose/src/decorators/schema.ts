import {decoratorTypeOf, StoreMerge, useDecorators} from "@tsed/core";
import {registerProvider} from "@tsed/di";
import {Property} from "@tsed/schema";
import {SchemaTypeOptions} from "mongoose";
import {MONGOOSE_SCHEMA} from "../constants";
import {MongooseSchemaOptions} from "../interfaces";
import {getSchema, getSchemaToken} from "../utils/createSchema";

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function Schema(options?: MongooseSchemaOptions): (target: any) => void;
export function Schema(definition: SchemaTypeOptions<any>): Function;
export function Schema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return (...parameters: any[]) => {
    switch (decoratorTypeOf(parameters)) {
      case "property":
        return useDecorators(Property(), StoreMerge(MONGOOSE_SCHEMA, options))(parameters[0], parameters[1], parameters[2]);

      case "class":
        const {token} = getSchemaToken(parameters[0], options);

        registerProvider({
          provide: token,
          deps: [],
          useFactory() {
            return getSchema(parameters[0], options as any);
          }
        });
        break;
    }
  };
}

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function MongooseSchema(options?: MongooseSchemaOptions): (target: any) => void;
export function MongooseSchema(definition: SchemaTypeOptions<any>): Function;
export function MongooseSchema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return Schema(options as any);
}
