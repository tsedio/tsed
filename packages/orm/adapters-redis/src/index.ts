/**
 * @file Automatically generated by @tsed/barrels.
 */
export * from "./adapters/OIDCRedisAdapter.js";
export * from "./adapters/RedisAdapter.js";
export * from "./domain/RedisAdaptersSettings.js";
