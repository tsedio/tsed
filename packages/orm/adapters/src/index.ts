export * from "./adapters/FileSyncAdapter";
export * from "./adapters/LowDbAdapter";
export * from "./adapters/MemoryAdapter";
export * from "./domain/Adapter";
export * from "./domain/AdaptersSettings";
export * from "./decorators/injectAdapter";
export * from "./decorators/indexed";
export * from "./services/Adapters";
