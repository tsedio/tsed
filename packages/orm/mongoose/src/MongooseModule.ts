import {PlatformCacheSettings, Configuration, OnInit} from "@tsed/common";
import {Module, OnDestroy} from "@tsed/di";
import {MONGOOSE_CONNECTIONS} from "./services/MongooseConnections";
import {MongooseService} from "./services/MongooseService";

/**
 * @ignore
 */
@Module({
  imports: [MONGOOSE_CONNECTIONS]
})
export class MongooseModule implements OnDestroy, OnInit {
  constructor(private mongooseService: MongooseService, @Configuration() private settings: Configuration) {
    // auto configure the cache manager when mongoose is used with @tsed/mongoose
    const cache = this.settings.get<PlatformCacheSettings>("cache");

    // istanbul ignore next
    if (cache?.mongoose) {
      cache.connection = this.mongooseService.get();
    }
  }

  $onInit(): void | Promise<any> {}

  $onDestroy(): Promise<any> | void {
    return this.mongooseService.closeConnections();
  }
}
