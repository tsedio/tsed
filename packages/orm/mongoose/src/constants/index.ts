/**
 *
 * @type {string}
 */
export const MONGOOSE_SCHEMA = "tsed:mongoose:schema";
/**
 *
 * @type {string}
 */
export const MONGOOSE_SCHEMA_OPTIONS = "tsed:mongoose:schema:options";
/**
 *
 * @type {string}
 */
export const MONGOOSE_MODEL = "tsed:mongoose:model";
/**
 *
 * @type {string}
 */
export const MONGOOSE_MODEL_NAME = "tsed:mongoose:model:name";
