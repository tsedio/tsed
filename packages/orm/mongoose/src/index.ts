export * from "./interfaces";
export * from "./constants";
export * from "./services/MongooseService";
export * from "./MongooseModule";
export * from "./utils";
export * from "./decorators";
