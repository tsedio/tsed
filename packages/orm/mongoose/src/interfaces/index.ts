import {MongooseConnectionOptions} from "./MongooseConnectionOptions";

declare global {
  namespace TsED {
    interface Configuration {
      mongoose: Omit<MongooseConnectionOptions, "id"> | MongooseConnectionOptions[];
    }
  }
}

export * from "./MongooseConnectionOptions";
export * from "./MongooseDocument";
export * from "./MongooseModel";
export * from "./MongooseModelOptions";
export * from "./MongooseSchemaOptions";
export * from "./MongooseVirtualRefOptions";
