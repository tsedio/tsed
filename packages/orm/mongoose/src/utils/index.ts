export * from "./createSchema";
export * from "./createModel";
