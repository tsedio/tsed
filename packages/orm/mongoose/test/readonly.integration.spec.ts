import {Inject, Injectable, PlatformTest} from "@tsed/common";
import {getJsonSchema, Groups, Name, Property, ReadOnly, Required} from "@tsed/schema";
import {TestMongooseContext} from "@tsed/testing-mongoose";
import {Immutable, Model, MongooseModel, ObjectID, SchemaIgnore} from "../src/index";

class BaseModel {
  @ObjectID("id")
  @Required()
  @Groups("!create")
  _id: string;

  @Required()
  @Immutable()
  @ReadOnly()
  @SchemaIgnore()
  @Name("created_at")
  @Groups("!create", "!update")
  createdAt: number = Date.now();

  @Required()
  @ReadOnly()
  @SchemaIgnore()
  @Name("updated_at")
  @Groups("!create", "!update")
  updatedAt: number = Date.now();
}

@Model()
class DataSourceModel extends BaseModel {
  @Property()
  @ReadOnly()
  get test() {
    return "test";
  }
}

@Injectable()
class MyService {
  @Inject(DataSourceModel)
  model: MongooseModel<DataSourceModel>;

  save() {
    const instance = new this.model({});
    return instance.save();
  }
}

describe("Mongoose: ReadOnly", () => {
  beforeEach(TestMongooseContext.create);
  afterEach(TestMongooseContext.reset);
  it("should generate json schema", () => {
    const jsonSchema = getJsonSchema(DataSourceModel);

    expect(jsonSchema).toMatchSnapshot();
  });

  it("should generate model", async () => {
    const service = PlatformTest.get<MyService>(MyService);
    await service.save();

    const list = await service.model.find().exec();
    expect(list.length).toEqual(1);
    expect(list[0].toObject().test).toBeUndefined();
    expect(list[0].toClass().test).toEqual("test");
  });
});
