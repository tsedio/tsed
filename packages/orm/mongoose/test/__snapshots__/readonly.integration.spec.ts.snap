// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mongoose: ReadOnly > should generate json schema 1`] = `
{
  "properties": {
    "created_at": {
      "readOnly": true,
      "type": "number",
    },
    "id": {
      "description": "An ObjectID",
      "examples": [
        "5ce7ad3028890bd71749d477",
      ],
      "minLength": 1,
      "pattern": "^[0-9a-fA-F]{24}$",
      "type": "string",
    },
    "test": {
      "readOnly": true,
      "type": "object",
    },
    "updated_at": {
      "readOnly": true,
      "type": "number",
    },
  },
  "required": [
    "id",
    "created_at",
    "updated_at",
  ],
  "type": "object",
}
`;
